src/arch/src/aarch64/cache_info.rs

```diff
@@ -318,24 +356,208 @@ mod tests {
+        assert_eq!("shared cpu map, size, number of sets", err_string);
+
+        err_string = "".to_string();
+        writeln_special(&coherency_dir, "Instruction");
+        let res = cache.populate(0, &mut err_string);
+        assert_eq!(
+            format!("{}", res.unwrap_err()),
+            "Invalid cache configuration found for coherency_line_size: invalid digit found in string"
+        );
+    }
+
+    #[test]
+    fn test_cache_size() {
+        let cache_dir = TempDir::new_with_prefix("/tmp/cachedir").unwrap();
+        let mut cache: CacheInfo = CacheInfo::default();
+        cache.cache_dir = PathBuf::from(cache_dir.as_path());
+        let index_dir = min_viable_cache_setup(&mut cache);
+        let mut err_string = String::new();
+
+        let mut size_dir = index_dir;
+        size_dir.push("size");
+        writeln_special(&size_dir, "64K");
+        assert!(cache.populate(0, &mut err_string).is_ok());
+        assert_eq!(
+            "shared cpu map, coherency line size, number of sets",
+            err_string
+        );
+
+        err_string = "".to_string();
+        writeln_special(&size_dir, "64");
+        let res = cache.populate(0, &mut err_string);
+        assert_eq!(
+            format!("{}", res.unwrap_err()),
+            "Invalid cache configuration found for size: 64"
+        );
+
+        writeln_special(&size_dir, "64Z");
+        let res = cache.populate(0, &mut err_string);
+        assert_eq!(
+            format!("{}", res.unwrap_err()),
+            "Invalid cache configuration found for size: 64Z"
+        );
+    }
+
+    #[test]
+    fn test_cache_no_sets() {
+        let cache_dir = TempDir::new_with_prefix("/tmp/cachedir").unwrap();
+        let mut cache: CacheInfo = CacheInfo::default();
+        cache.cache_dir = PathBuf::from(cache_dir.as_path());
+        let index_dir = min_viable_cache_setup(&mut cache);
+
+        let mut sets_dir = index_dir;
+        sets_dir.push("number_of_sets");
+        writeln_special(&sets_dir, "64");
+        let mut err_string = String::new();
+        assert!(cache.populate(0, &mut err_string).is_ok());
+        assert_eq!("shared cpu map, coherency line size, size", err_string);
+
+        err_string = "".to_string();
+        writeln_special(&sets_dir, "64K");
+        let res = cache.populate(0, &mut err_string);
+        assert_eq!("shared cpu map, coherency line size, size", err_string);
+        assert_eq!(
+            format!("{}", res.unwrap_err()),
+            "Invalid cache configuration found for number_of_sets: invalid digit found in string"
+        );
+    }
+
+    #[test]
+    fn test_sysfs_read_caches() {
+        let mut l1_caches: Vec<CacheInfo> = Vec::new();
+        let mut non_l1_caches: Vec<CacheInfo> = Vec::new();
+        // We use sysfs for extracting the cache information.
+        assert!(sysfs_read_caches(&mut l1_caches, &mut non_l1_caches).is_ok());
+        assert_eq!(l1_caches.len(), 2);
```

Comment:
Are these assumptions we're willing to make? These depend on the cache topology of the host, right?

If somebody is running these unit tests on another machine that does not have these, they will run into errors I'm assuming.

Can we somehow mock the sysfs for this test?

Reply:
Indeed, these asserts depend on the host configuration and they might fail when run on other machines. I removed them.
Mocking a sysfs is not quite obvious given that `sysfs_read_caches` (the function that we are testing) constructs the `cache` object inside it and so I do not have access to the `cache_dir` field.

Reply:
I see. However, can't we mock the `cache_dir` by implementing the `Default` trait in the test mod with a custom path?